contracts/staking/StakingManager.sol

```diff
@@ -0,0 +1,447 @@
+        uint256 maximumStakeAmount;
+        uint64 minimumStakeDuration;
+        uint8 maximumHourlyChurn;
+        IRewardCalculator rewardCalculator;
+    }
+
+    /**
+     * @notice Warp precompile used for sending and receiving Warp messages.
+     */
+    IWarpMessenger public constant WARP_MESSENGER =
+        IWarpMessenger(0x0200000000000000000000000000000000000005);
+
+    function initialize(StakingManagerSettings calldata settings) public initializer {
+        __StakingManager_init(settings);
+    }
+
+    // solhint-disable-next-line func-name-mixedcase
+    function __StakingManager_init(StakingManagerSettings calldata settings)
+        internal
+        onlyInitializing
+    {
+        __ReentrancyGuard_init();
+        __Context_init();
+        __StakingManager_init_unchained(settings);
+    }
+
+    // solhint-disable-next-line func-name-mixedcase
+    function __StakingManager_init_unchained(StakingManagerSettings calldata settings)
+        internal
+        onlyInitializing
+    {
+        StakingManagerStorage storage $ = _getStakingManagerStorage();
+        $._pChainBlockchainID = settings.pChainBlockchainID;
+        $._subnetID = settings.subnetID;
+        $._minimumStakeAmount = settings.minimumStakeAmount;
+        $._maximumStakeAmount = settings.maximumStakeAmount;
+        $._minimumStakeDuration = settings.minimumStakeDuration;
+        $._maximumHourlyChurn = settings.maximumHourlyChurn;
+        $._rewardCalculator = settings.rewardCalculator;
+    }
+
+    /**
+     * @notice Begins the validator registration process. Locks the provided native asset in the contract as the stake.
+     * @param nodeID The node ID of the validator being registered.
+     * @param registrationExpiry The time at which the reigistration is no longer valid on the P-Chain.
+     * @param signature The raw bytes of the Ed25519 signature over the concatenated bytes of
+     * [subnetID]+[nodeID]+[blsPublicKey]+[weight]+[balance]+[expiry]. This signature must correspond to the Ed25519
+     * public key that is used for the nodeID. This approach prevents NodeIDs from being unwillingly added to Subnets.
+     * balance is the minimum initial $nAVAX balance that must be attached to the validator serialized as a uint64.
+     * The signature field will be validated by the P-Chain. Implementations may choose to validate that the signature
+     * field is well-formed but it is not required.
+     */
+    function _initializeValidatorRegistration(
+        bytes32 nodeID,
+        uint256 value,
+        uint64 registrationExpiry,
+        bytes memory signature
+    ) internal nonReentrant returns (bytes32) {
+        StakingManagerStorage storage $ = _getStakingManagerStorage();
+
+        // Ensure the registration expiry is in a valid range.
+        require(
+            registrationExpiry > block.timestamp && block.timestamp + 2 days > registrationExpiry,
+            "StakingManager: Invalid registration expiry"
```

Comment:
[Solidity 0.8.26 allows you to use custom errors in `require()`](https://soliditylang.org/blog/2024/05/21/solidity-0.8.26-release-announcement/#:~:text=Custom%20errors%20support%20in%20require). They result in smaller contract sizes and are IMO much better for tests because you can assert the exact reason for a failure without risk of introducing a [change-detector test](https://testing.googleblog.com/2015/01/testing-on-toilet-change-detector-tests.html) if you modify the error string. If you want to stick to 0.8.25 then I'd recommend switching to `if (!cond) { revert SpecificError() }` where you originally had `require(cond, ...)`.

Reply:
Our versioning policy is to use the second-most-recent release to hedge against potential issues in the latest release. That said, we've struggled to choose between revert strings and custom errors in the past, so getting the best of both worlds will be a welcome change.